tensorflow_model_optimization/python/core/sparsity/keras/test_utils.py

```diff
@@ -173,4 +173,4 @@ def assert_model_sparsity(test_case, sparsity, model):
     if isinstance(layer, pruning_wrapper.PruneLowMagnitude):
       for weight in layer.layer.get_prunable_weights():
         test_case.assertAllClose(
-            sparsity, _get_sparsity(tf.keras.backend.get_value(weight)))
+            sparsity, _get_sparsity(tf.keras.backend.get_value(weight)), rtol=1e-4, atol=1e-4)
```

Comment:
Which tests were sporadically failing that required this tolerance change?

Could you only modify the tolerances for that specific set (e.g. make assert_model_sparsity have new rtol and atol args)? Then leave a TODO(tfmot) to investigate with any details you know that may help with fixing in the future.

Furthermore, if the failures only happen on some form of eager or graph mode, that would help too.

Thanks! Keeping the tests as precise as possible will help when pruning migrates to a new internal abstraction that will eventually be used by weight clustering and other techniques.

Reply:
Done, added to TestPrunesModel where the failure happens (happens both in eager and non-eager and in funcmodel and sequential). However, I did not add a TODO comment as this is not bug, but just a consequence of the very low default tolerance sometimes not achieved even when the sparsity is correct. To demonstrate this, the test output for one of the failing cases is below:

`Not equal to tolerance rtol=1e-06, atol=1e-06
Mismatched value: a is different from b.
not close lhs = 0.0
not close rhs = 1.9531249999982236e-05
not close dif = 1.9531249999982236e-05
not close tol = 1.0000195312499998e-06
dtype = float64, shape = ()
Mismatch: 100%
Max absolute difference: 1.953125e-05
Max relative difference: 1.
 x: array(0.)
 y: array(1.953125e-05)`

Reply:
Thanks. I suspect the problem for the inconsistency is the following: if there are 1000 weights in layer and some threshold for pruning, you might have 501 weights that lie at the 50% threshold. If the implementation prunes them all, then you no longer have a perfect 50% sparsity. 1e-05 could be that extra weight(s). 

https://github.com/tensorflow/model-optimization/issues/215 would be relevant to this then. This cross link should suffice for now.

Reply:
Yes, I suspect this is the problem. Thanks for approving!